File: tests/serviceplans.fs
module regiondeployer.tests.serviceplans

open Xunit
open regiondeployer.cloudactions.buggedlinuxapp
open regiondeployer.cloudactions.serviceplan
open tests
open storage
open Xunit.Abstractions
open System
open Microsoft.Azure.Management.AppService.Fluent
open System
open regiondeployer.clouderror
open regiondeployer

type public when_a_linux_service_plan_is_created(inputlogger:ITestOutputHelper) =
    let storagetestsfield = new when_a_storage_account_is_created(inputlogger)
    let storagecontextsfield = storagetestsfield.it_succeeds()

    do
        configadapter.initialise();

    member public x.storagetests = storagetestsfield
    member public x.storagecontext = storagecontextsfield

    interface IDisposable with member this.Dispose() =
        (storagetestsfield :> IDisposable).Dispose()

    [<Fact>]
    member public this.it_succeeds() : IAppServicePlan =
        try
            let testcontext:IAppServicePlan = createlinuxserviceplan(
                storagetestsfield.groupcontext, 
                inputlogger.WriteLine
            )
            testcontext

        with 
        | error ->
            printclouderror(error, inputlogger.WriteLine)
            raiseclouderror error
            null
        
type public when_a_bugged_linux_app_fix_is_applied(inputlogger:ITestOutputHelper) = 
    let serviceplantests = new when_a_linux_service_plan_is_created(inputlogger)
    let serverplancontext = serviceplantests.it_succeeds()

    let is = true

    interface IDisposable with member this.Dispose() =
        (serviceplantests :> IDisposable).Dispose()

    [<Fact>]
    member public this.it_succeeds() : unit =
        try
            let resourcegroupname = serviceplantests.storagetests.groupcontext
            let storagegroup = serviceplantests.storagecontext
            fixlinuxappbug(resourcegroupname, inputlogger.WriteLine, storagegroup)
            ()

        with 
        | error ->
            printclouderror(error, inputlogger.WriteLine)
            raiseclouderror error
            ()

File: tests/ServicePlans.cs
using Xunit;
using Xunit.Abstractions;
using Microsoft.Azure.Management.AppService.Fluent;
using System;

namespace Tests
{
    public class WhenALinuxServicePlanIsCreated : IDisposable
    {
        private readonly ITestOutputHelper _inputLogger;
        private readonly WhenAStorageAccountIsCreated _storageTestsField;
        private readonly IAppServicePlan _storageContextsField;

        public WhenALinuxServicePlanIsCreated(ITestOutputHelper inputLogger)
        {
            _inputLogger = inputLogger;
            _storageTestsField = new WhenAStorageAccountIsCreated(inputLogger);
            _storageContextsField = _storageTestsField.ItSucceeds();

            ConfigAdapter.Initialise();
        }

        public WhenAStorageAccountIsCreated StorageTests => _storageTestsField;
        public IAppServicePlan StorageContext => _storageContextsField;

        public void Dispose()
        {
            (_storageTestsField as IDisposable)?.Dispose();
        }

        [Fact]
        public IAppServicePlan ItSucceeds()
        {
            try
            {
                var testContext = CreateLinuxServicePlan(
                    _storageTestsField.GroupContext,
                    _inputLogger.WriteLine
                );
                return testContext;
            }
            catch (Exception error)
            {
                PrintCloudError(error, _inputLogger.WriteLine);
                RaiseCloudError(error);
                return null;
            }
        }
    }

    public class WhenABuggedLinuxAppFixIsApplied : IDisposable
    {
        private readonly WhenALinuxServicePlanIsCreated _servicePlanTests;
        private readonly IAppServicePlan _serverPlanContext;

        public WhenABuggedLinuxAppFixIsApplied(ITestOutputHelper inputLogger)
        {
            _servicePlanTests = new WhenALinuxServicePlanIsCreated(inputLogger);
            _serverPlanContext = _servicePlanTests.ItSucceeds();
        }

        public void Dispose()
        {
            (_servicePlanTests as IDisposable)?.Dispose();
        }

        [Fact]
        public void ItSucceeds()
        {
            try
            {
                var resourceGroupName = _servicePlanTests.StorageTests.GroupContext;
                var storageGroup = _servicePlanTests.StorageContext;
                FixLinuxAppBug(resourceGroupName, _inputLogger.WriteLine, storageGroup);
            }
            catch (Exception error)
            {
                PrintCloudError(error, _inputLogger.WriteLine);
                RaiseCloudError(error);
            }
        }
    }
}